vignettes/Rcpp-modules.Rmd

```diff
@@ -1059,74 +1136,183 @@ RCPP_MODULE(mod_vec) {
                   mustStart = TRUE)
 vec <- mod_vec$vec
-# and that is not needed in a package
-# setup as e.g. one created
-# via Rcpp.package.skeleton(..., module=TRUE)
 v <- new(vec)
 v$reserve(50L)
 v$assign(1:10)
 v$push_back(10)
 v$size()
+v$resize(30L)
 v$capacity()
 v[[ 0L ]]
 v$as.vector()
 ```
 
-# Using modules in other packages {#sec:package}
+<!-- CHECK-PR: sourceCpp is pretty convenient, OK to add as new section? -->
+<!-- If not, the content of yada.cpp should be shown in the package section -->
+## Loading modules via sourceCpp {#sec:modules-sourceCpp}
+
+As an alternative to the explicit creation of a `Module` object using the
+\pkg{inline} package via `cxxfunction` and `getDynLib`, it is possible to use
+the `sourceCpp` function, accepting \proglang{C++} source code as either a
+`.cpp` file or a character string. The main differences with this approach are:
 
-## Namespace import/export
+- The `Rcpp.h` header file must be explicitly included.
+- The content of the module (\proglang{C++} functions and classes) are
+implicitly exposed and made available to \proglang{R} as individual objects, as
+opposed to being accessed from a `Module` object with the `$` extractor.
 
-### Import all functions and classes
+Note that this is similar to exposing modules in \proglang{R} packages using
+`loadModule`, described in Section \ref{sec:package-namespace-loadModule} below.
 
-When using \pkg{Rcpp} modules in a packages, the client package needs to
+As an example, consider a file called `yada.cpp` containing the following
+\proglang{C++} code:
+
+```{Rcpp eval=params$eval}
+#include <Rcpp.h>
+std::string hello() {
+    return "hello";
+}
+void bla() {
+    Rprintf("hello\\n");
+}
+void bla2( int x, double y) {
+    Rprintf("hello (x = %d, y = %5.2f)\\n", x, y);
+}
+
+class World {
+public:
+    World() : msg("hello") {}
+    void set(std::string msg) { this->msg = msg; }
+    std::string greet() { return msg; }
+
+private:
+    std::string msg;
+};
+
+RCPP_MODULE(yada){
+    using namespace Rcpp;
+
+    function("hello" , &hello);
+    function("bla"   , &bla);
+    function("bla2"  , &bla2);
+
+    class_<World>("World")
+    .constructor()
+    .method("greet", &World::greet)
+    .method("set",   &World::set)
+    ;
+}
+```
+
+```{r}
+sourceCpp('yada.cpp')
+```
+
+\proglang{C++} functions `hello`, `bla`, `bla2` and class `World` will be
+readily available in \proglang{R}:
+
+```{r, eval=params$eval, results=params$results}
+hello()
+bla()
+bla2(42, 0.42)
+w <- new(World)
+w$greet()
+w$set("hohoho")
+w$greet()
+```
+
+
+# Using modules in other packages {#sec:package}
+
+## Namespace import
+
+When using \textsl{Rcpp modules} in a packages, the client package needs to
 import \pkg{Rcpp}'s namespace. This is achieved by adding the
 following line to the `NAMESPACE` file.
 
+<!-- CHECK-PR: is this chunk needed at all? -->
 ```{r, echo=FALSE,eval=TRUE}
 options( prompt = " ", continue = " " )
 ```
 
-```{r, eval=FALSE}
+```{r}
```

Comment:
I feel like there was also a need to add:

```r
import(methods)
```

Reply:
I double-checked using the package produced by `Rcpp::Rcpp.package.skeleton()`, removed the DESCRIPTION and NAMESPACE imports of `methods` and things seem to still work.

Reply:
Thanks for checking that.  I will merge this PR but that whole aspect may be worth looking into in another ticket.  This changed over the years, and maybe describing just one simple (working) setup is best.